### PR TITLE
initial implementation of complevels

### PR DESCRIPTION
**Pull request by rfomin**
I am sure this could be better organized.

### CODE REVIEW

File: Source/g_game.c [hunk 1]
```diff
@@ -2226,6 +2228,32 @@ static int G_GetHelpers(void)
...
+
+static int G_GetNamedComplevel (const char *arg)
+{
+  int i;
+
+  const struct {
+    int level;
+    const char *const name;
+  } named_complevel[] = {
+    {109, "vanilla"},
```
Comment (fabiangreffrath) [inline Source/g_game.c:2241]: We should also add "2", "3", "4", "9" and "11" as aliases as well as the IWAD names. And we also have the concept of `gameversion` which should be taken into consideration in the "vanilla" case. Should choice of `-complevel` override `-gameversion` here if vice versa?

File: Source/g_game.c [hunk 2]
```diff
@@ -2286,13 +2314,45 @@ void G_ReloadDefaults(void)
...
+    int i = M_CheckParm("-complevel");
+    if (i && (1+i) < myargc) {
+      int l = G_GetNamedComplevel(myargv[i+1]);
+      if (l >= -1) complevel = l;
+    }
+  }
+  if (complevel == -1)
+    complevel = MBFVERSION;
+
+  demo_version = complevel;//MBFVERSION;     // killough 7/19/98: use this version's id
```
Comment (fabiangreffrath) [inline Source/g_game.c:2328]: Please remove the double comment.

File: Source/g_game.c [hunk 3]
```diff
@@ -2678,6 +2738,9 @@ void G_BeginRecording(void)
 
   demo_p = demobuffer;
 
+  if (complevel == MBFVERSION)
+  {
+
   *demo_p++ = MBFVERSION;
```
Comment (fabiangreffrath) [inline Source/g_game.c:2744]: Excess blank line.

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Looks promising, thanks!

File: Source/g_game.c (hunk 1 above)
Comment (rfomin) [reply Source/g_game.c]: > Should choice of `-complevel` override `-gameversion`

I think it should. Did it in https://github.com/fabiangreffrath/woof/pull/157/commits/2ee34ac10ce5f64acbf47cfd56c869cd2ea4c1fc

**Comment by fabiangreffrath** [comment]
This is looking good already. 

Some scenarios to consider:
- If I start the game without any complevel parameter, will DEMO1 get played back with Vanilla compatibility and if I start a new game after that, will this have proper MBF compatibility?
- If I start the game with `-complevel mbf` and wait for the demo reel, which complevel will DEMO1 get played back with and if I start a new game after DEMO1, which complevel will this have?
- What if I play back a Vanilla demo with `-playdemo vanilla.lmp -complevel mbf`?
- Vice versa, what if I play back a MBF demo with `-playdemo mbf.lmp -complevel vanilla`?
- If I start the game with `-gameversion final -complevel vanilla` will this apply to DEMO1?
- etc...

**Comment by rfomin** [comment]
In general, the demo sets `demo_version` from the demo header and ignores complevel. But `-complevel 2, 3, 4` and `-gameversion` affect both demo playback and gameplay.

> * If I start the game without any complevel parameter, will DEMO1 get played back with Vanilla compatibility and if I start a new game after that, will this have proper MBF compatibility?

Yes, works fine.

> * If I start the game with `-complevel mbf` and wait for the demo reel, which complevel will DEMO1 get played back with and if I start a new game after DEMO1, which complevel will this have?

DEMO1 will be vanilla and gameplay will be mbf.

> * What if I play back a Vanilla demo with `-playdemo vanilla.lmp -complevel mbf`?

Demo will be vanilla.

> * Vice versa, what if I play back a MBF demo with `-playdemo mbf.lmp -complevel vanilla`?

Demo will be mbf and gameplay vanilla.

> * If I start the game with `-gameversion final -complevel vanilla` will this apply to DEMO1?

Yes, it will apply.

**Comment by fabiangreffrath** [comment]
Sounds like it's working as supposed to be, then.

**Comment by fabiangreffrath** [comment]
One last thing: to avoid another string comparison, could you turn the third element of the `named_complevel[]` array into an int and replace `"auto"` with `-1` and the gameversion with their numerical values and only set gameversion for values >0.

**Comment by fabiangreffrath** [comment]
Thank you!

**Pull request merged**